datasets/flwr_datasets/partitioner/inner_dirichlet_partitioner.py

```diff
@@ -0,0 +1,271 @@
+# Copyright 2023 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""InnerDirichlet partitioner."""
+import warnings
+from typing import Dict, List, Optional, Union
+
+import numpy as np
+
+import datasets
+from flwr_datasets.common.typing import NDArrayFloat, NDArrayInt
+from flwr_datasets.partitioner.partitioner import Partitioner
+
+
+class InnerDirichletPartitioner(Partitioner):  # pylint: disable=R0902
+    """Partitioner based on Dirichlet distribution.
+
+    Implementation based on Federated Learning Based on Dynamic Regularization
+    https://arxiv.org/abs/2111.04263.
```

Comment:
I wonder if we should format slightly the title of the paper. Maybe having it between quotes? Or something like:

```
Implementation based on: Federated Learning Based on Dynamic Regularization (https://arxiv.org/abs/2111.04263).
```

Reply:
It's consistent with how I describe other partitioners. Let's agree on the format here, and then I'll change it here and open another PR, changing it in other places.
Implementation based on [description in]: TITLE (URL). 
[description in] would be optional and used only if we want to emphasize that no code was available.
And I'd propose it should be the last paragraph in the docs (I was inconsistent regarding the placement). The top docs should explain how the algorithm works.
Wdyt?

Reply:
`TITLE (URL)` looks ok. I agree placing this at the bottom is better.